storey/steps/sample.py

```diff
@@ -42,31 +49,59 @@ def __init__(
         self._window_size = window_size
         self._emit_period = emit_period
         self._emit_before_termination = emit_before_termination
-
+        self._per_key_count = dict()
         self._count = 0
         self._last_event = None
+        self._extract_key: Callable[[Event], str] = self._create_key_extractor(key)
+
+    @staticmethod
+    def _create_key_extractor(key) -> Callable:
+        if key is None:
+            return lambda event: None
+        elif callable(key):
+            return key
+        elif isinstance(key, str):
+            if key == '$key':
+                return lambda event: event.key
+            else:
+                return lambda event: event.body[key]
+        else:
+            raise ValueError(f'Unsupported key type {type(key)}')
 
     async def _do(self, event):
         if event is _termination_obj:
             if self._last_event is not None:
                 await self._do_downstream(self._last_event)
             return await self._do_downstream(_termination_obj)
         else:
-            self._count += 1
+            key = self._extract_key(event)
+
+            if key is not None:
+                if key not in self._per_key_count:
+                    self._per_key_count[key] = 1
+                else:
+                    self._per_key_count[key] += 1
+                count = self._per_key_count[key]
+            else:
+                self._count += 1
+                count = self._count
 
             if self._emit_before_termination:
                 self._last_event = event
 
-            if self._should_emit():
+            if self._should_emit(count):
                 self._last_event = None
                 await self._do_downstream(event)
```

Comment:
Not new in this PR, but I'm worried that this could throw and error and we won't reset the count, causing it to increment infinitely going forward, and never emitting again.

Reply:
This wasn't as much as a problem in the past, but now the default behaviour in EmitSource is to supress errors, and there is the concept of recovery steps (which may be upstream of the sampler).